### ISSUE TITLE
New AgentMacroReference Event Type

### ISSUE THREAD

**Issue by matthewtckr** [cloudbees/zendesk-java-client#126]
Came across this in our ETL process, it appears to be a new Zendesk Event type that's not implemented, and apparently no documentation from Zendesk.

Sample Event from a ticket audit entry:

```
{
            "id": 121204214906,
            "type": "AgentMacroReference",
            "via": {
                "channel": "web",
                "source": {
                    "from": {},
                    "to": {},
                    "rel": null
                }
            },
            "macro_title": "Workflow Stage: Problem Definition",
            "macro_id": "49985043"
        }
```

Sample error for unrecognized event (Zendesk Java Client 0.4.0):

```
Unexpected error
org.zendesk.client.v2.ZendeskException: java.lang.IllegalArgumentException: Could not resolve type id 'AgentMacroReference' into a subtype of [simple type, class org.zendesk.client.v2.model.events.Event]
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Audit["events"])
    at org.zendesk.client.v2.Zendesk.complete(Zendesk.java:1873)
    at org.zendesk.client.v2.Zendesk.access$1300(Zendesk.java:83)
    at org.zendesk.client.v2.Zendesk$PagedIterable$PagedIterator.hasNext(Zendesk.java:2055)
    at org.pentaho.di.trans.steps.zendesk.ZendeskInputTicketAudit.processRow(ZendeskInputTicketAudit.java:115)
    at org.pentaho.di.trans.step.RunThread.run(RunThread.java:62)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Could not resolve type id 'AgentMacroReference' into a subtype of [simple type, class org.zendesk.client.v2.model.events.Event]
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Audit["events"])
    at com.ning.http.client.providers.netty.future.NettyResponseFuture.done(NettyResponseFuture.java:220)
    at com.ning.http.client.providers.netty.handler.HttpProtocol.finishUpdate(HttpProtocol.java:194)
    at com.ning.http.client.providers.netty.handler.HttpProtocol.handleChunk(HttpProtocol.java:451)
    at com.ning.http.client.providers.netty.handler.HttpProtocol.handle(HttpProtocol.java:474)
    at com.ning.http.client.providers.netty.handler.Processor.messageReceived(Processor.java:88)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
    at org.jboss.netty.handler.stream.ChunkedWriteHandler.handleUpstream(ChunkedWriteHandler.java:142)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
    at org.jboss.netty.handler.codec.http.HttpContentDecoder.messageReceived(HttpContentDecoder.java:135)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:452)
    at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536)
    at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
    at org.jboss.netty.handler.codec.http.HttpClientCodec.handleUpstream(HttpClientCodec.java:92)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:310)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
    at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
    at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
    at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
    at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
    at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    ... 1 more
Caused by: java.lang.IllegalArgumentException: Could not resolve type id 'AgentMacroReference' into a subtype of [simple type, class org.zendesk.client.v2.model.events.Event]
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Audit["events"])
    at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:2518)
    at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:2445)
    at org.zendesk.client.v2.Zendesk$PagedAsyncListCompletionHandler.onCompleted(Zendesk.java:1697)
    at org.zendesk.client.v2.Zendesk$PagedAsyncListCompletionHandler.onCompleted(Zendesk.java:1681)
    at com.ning.http.client.AsyncCompletionHandler.onCompleted(AsyncCompletionHandler.java:55)
    at com.ning.http.client.providers.netty.future.NettyResponseFuture.getContent(NettyResponseFuture.java:177)
    at com.ning.http.client.providers.netty.future.NettyResponseFuture.done(NettyResponseFuture.java:214)
    ... 41 more
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Could not resolve type id 'AgentMacroReference' into a subtype of [simple type, class org.zendesk.client.v2.model.events.Event]
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Audit["events"])
    at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:164)
    at com.fasterxml.jackson.databind.DeserializationContext.unknownTypeException(DeserializationContext.java:701)
    at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._findDeserializer(TypeDeserializerBase.java:155)
    at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:98)
    at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:82)
    at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:107)
    at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:228)
    at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:203)
    at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:23)
    at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:375)
    at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:98)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:308)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:121)
    at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:2513)
    ... 47 more
```

**Comment by matthewtckr** [comment]
Here's the response from the Zendesk Support case:

> Thanks again for your patience while I got more info from our Product team.
> 
> They let me know that we do not describe every event type or attribute in our documentation. We may occasionally add attributes and event types, and so it's important for your integration to be flexible enough to encounter account unknown event and attribute types without breaking.
> 
> That aside, we have asked our documentation team to take a look at this event type for possible addition to our existing public documentation. We're also asking them to add a note for developers to be aware of what I mentioned above regarding unknown and additional event types.
> 
> One other thing I wanted to point out is that the Java API Client you linked me to is not developed by our officially supported by Zendesk. It looks like the developer is called Cloudbees. If it is proving to be fragile in the way you described, you can either fork it on Github to modify it for your own purposes, or you might want to contact them directly to let them know it is not handling these exceptions very well.
> 
> I'm sorry if there's not more I can do for you here, but I did want to let you know again that we take the issues you raised seriously, and that we'll be reviewing our documentation with those issues in mind!

**Comment by SlalomJonMeran** [comment]
Does anyone know when the next release with this fix will make its way into Maven?

**Comment by matthewtckr** [comment]
Closing, as it was addressed as part of the 0.5.2 release which was made available today.
